geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java

```diff
@@ -127,50 +127,139 @@ private int getArrayIndex(int listIndex) {
+      removed = new RemoveElementsByIndex(newVersion);
+      removed.add(0);
+    }
     storeChanges(region, key, removed);
     return popped;
   }
 
+  public synchronized byte[] removeFirstElement() {
+    return elementList.removeFirst();
+  }
+
+  public synchronized byte[] removeLastElement() {
+    return elementList.removeLast();
+  }
+
   /**
-   * @return the number of elements in the list
+   * @param elementsToAdd elements to add to this list; NOTE this list may be modified by this call
+   * @param region the region this instance is stored in
+   * @param key the name of the list to add to
+   * @param onlyIfExists if true then the elements should only be added if the key already exists
+   *        and holds a list, otherwise no operation is performed.
+   * @return the length of the list after the operation
    */
-  public int llen() {
+  public long lpush(List<byte[]> elementsToAdd, Region<RedisKey, RedisData> region, RedisKey key,
+      final boolean onlyIfExists) {
+    byte newVersion;
+    AddByteArrays addByteArrays;
+    synchronized (this) {
+      newVersion = incrementAndGetVersion();
+      elementsPush(elementsToAdd);
+      addByteArrays = new AddByteArrays(newVersion, elementsToAdd);
+    }
+    storeChanges(region, key, addByteArrays);
     return elementList.size();
   }
 
+  /**
+   * @param elementsToAdd elements to add to this list; NOTE this list may be modified by this call
+   * @param region the region this instance is stored in
+   * @param key the name of the list to add to
+   * @return the number of elements actually added
+   */
+  public long lpush(List<byte[]> elementsToAdd, Region<RedisKey, RedisData> region, RedisKey key) {
+    byte newVersion;
+    AddByteArrays addByteArrays;
+    synchronized (this) {
+      newVersion = incrementAndGetVersion();
+      elementsPush(elementsToAdd);
+      addByteArrays = new AddByteArrays(newVersion, elementsToAdd);
+    }
+    storeChanges(region, key, addByteArrays);
+    return elementList.size();
+  }
+
+  public byte[] ltrim(long start, long end, Region<RedisKey, RedisData> region,
+      RedisKey key) {
+    byte newVersion;
+    int length = elementList.size();
+    int boundedStart = getBoundedStartIndex(start, length);
+    int boundedEnd = getBoundedEndIndex(end, length);
+    List<Integer> removed = new ArrayList<>();
+    RemoveElementsByIndex removeElementsByIndex;
+
+    synchronized (this) {
+      if (boundedStart > boundedEnd || boundedStart == length) {
+        // Remove everything
+        for (int i = length - 1; i >= 0; i--) {
+          removed.add(i);
+        }
+      } else {
+        // Remove any elements after boundedEnd
+        for (int i = length - 1; i > boundedEnd; i--) {
+          removed.add(i);
+        }
+
+        // Remove any elements before boundedStart
+        for (int i = boundedStart - 1; i >= 0; i--) {
+          removed.add(i);
+        }
+      }
+
+      if (removed.size() > 0) {
```

Comment:
no need for this size check. The for statement in elementsRemove will do nothing if remove is empty

Reply:
Mooted by redesign.